File: adapters/cursor.py
"""Cursor (VS Code clone) theme adapter."""

import json
from pathlib import Path
from .base import ThemeAdapter

DEFAULT_SETTINGS = Path.home() / ".config" / "Cursor" / "User" / "settings.json"
DEFAULT_THEMES = [
    "Default Dark+", "Default Dark Modern", "Default Light Modern",
    "Default Light+", "Quiet Light", "Default High Contrast",
]


class CursorAdapter(ThemeAdapter):
    name = "cursor"
    display_name = "Cursor"
    supports_preview = False

    @property
    def settings_path(self) -> Path:
        return Path(self.settings.get("settings_path", str(DEFAULT_SETTINGS)))

    def list_themes(self) -> list[str]:
        return self.settings.get("known_themes", list(DEFAULT_THEMES))

    def get_current(self) -> str | None:
        path = self.settings_path
        if not path.exists():
            return None
        data = json.loads(path.read_text())
        # Check for explicit colorTheme first
        if "workbench.colorTheme" in data:
            return data["workbench.colorTheme"]
        # Fall back to preferred dark/light theme
        if data.get("window.autoDetectColorScheme"):
            return data.get("workbench.preferredDarkColorTheme",
                            data.get("workbench.preferredLightColorTheme"))
        return None

    def commit(self, theme_name: str) -> bool:
        path = self.settings_path
        if not path.exists():
            return False
        data = json.loads(path.read_text())
        # Set the theme — if autoDetect is on, set the preferred variant
        if data.get("window.autoDetectColorScheme"):
            # Heuristic: "light" in name -> light theme, else dark
            is_light = any(w in theme_name.lower() for w in ("light", "quiet"))
            if is_light:
                data["workbench.preferredLightColorTheme"] = theme_name
            else:
                data["workbench.preferredDarkColorTheme"] = theme_name
        else:
            data["workbench.colorTheme"] = theme_name
        path.write_text(json.dumps(data, indent=4) + "\n")
        return True

    def describe(self, theme_name: str) -> str:
        current = self.get_current()
        marker = " (current)" if current == theme_name else ""
        return f"  {self.display_name}: {theme_name}{marker}"

File: internal/adapters/cursor.go
package adapters

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/joebhakim/themer/internal/config"
	"github.com/joebhakim/themer/internal/core"
	"github.com/tailscale/hujson"
)

type Cursor struct {
	cfg config.CursorConfig
}

func NewCursor(cfg config.CursorConfig) *Cursor {
	return &Cursor{cfg: cfg}
}

func (c *Cursor) Name() string {
	return "cursor"
}

func (c *Cursor) DisplayName() string {
	return "Cursor"
}

func (c *Cursor) Validate(context.Context) []core.Diagnostic {
	if _, err := os.Stat(c.cfg.SettingsPath); err != nil {
		return []core.Diagnostic{{
			Adapter: c.Name(),
			Level:   "warn",
			Message: fmt.Sprintf("settings path %s is unavailable", c.cfg.SettingsPath),
		}}
	}
	return nil
}

func (c *Cursor) ListThemes(context.Context) ([]string, error) {
	return append([]string(nil), c.cfg.KnownThemes...), nil
}

func (c *Cursor) Current(context.Context) (string, error) {
	settings, err := c.load()
	if err != nil {
		return "", err
	}
	if theme, ok := settings["workbench.colorTheme"].(string); ok && strings.TrimSpace(theme) != "" {
		return theme, nil
	}
	if enabled, ok := settings["window.autoDetectColorScheme"].(bool); ok && enabled {
		if theme, ok := settings["workbench.preferredDarkColorTheme"].(string); ok && theme != "" {
			return theme, nil
		}
		if theme, ok := settings["workbench.preferredLightColorTheme"].(string); ok && theme != "" {
			return theme, nil
		}
	}
	return "", nil
}

func (c *Cursor) Describe(context.Context, string) (core.ThemeDescription, error) {
	return core.ThemeDescription{
		Summary: "Cursor or VS Code theme stored in settings.json (JSONC supported).",
		Notes:   []string{"Apply-only adapter; theme is written back as formatted JSON."},
	}, nil
}

func (c *Cursor) PreviewStatus(context.Context) core.PreviewSupport {
	return core.PreviewSupport{Reason: "cursor adapter is apply-only"}
}

func (c *Cursor) Preview(context.Context, string) (func(context.Context) error, error) {
	return nil, fmt.Errorf("cursor does not support preview")
}

func (c *Cursor) Apply(ctx context.Context, theme string) error {
	settings, err := c.load()
	if err != nil {
		return err
	}
	if enabled, ok := settings["window.autoDetectColorScheme"].(bool); ok && enabled {
		name := strings.ToLower(theme)
		if strings.Contains(name, "light") || strings.Contains(name, "quiet") {
			settings["workbench.preferredLightColorTheme"] = theme
		} else {
			settings["workbench.preferredDarkColorTheme"] = theme
		}
	} else {
		settings["workbench.colorTheme"] = theme
	}
	data, err := json.MarshalIndent(settings, "", "  ")
	if err != nil {
		return err
	}
	return atomicWriteFile(c.cfg.SettingsPath, append(data, '\n'), 0o644)
}

func (c *Cursor) load() (map[string]any, error) {
	data, err := os.ReadFile(c.cfg.SettingsPath)
	if err != nil {
		return nil, err
	}
	value, err := hujson.Parse(data)
	if err != nil {
		return nil, err
	}
	value.Standardize()
	standardized := value.Pack()
	settings := map[string]any{}
	if err := json.Unmarshal(standardized, &settings); err != nil {
		return nil, err
	}
	return settings, nil
}
